Src/DependencyCollector/Shared/ServiceBusDiagnosticsListener.cs

```diff
@@ -0,0 +1,172 @@
+        // This way, UI could have special representation for all kinds of queues
+        // Not knowing particular service
+        private const string DependencyType = "Queue.ServiceBus";
+
+        private readonly TelemetryClient telemetryClient;
+
+        // Every fetcher is unique for event payload and particular property into this payload
+        // when we first receive an event and require particular property, we cache fetcher for it
+        // There are just a few (~10) events we can receive and each will a have a few payload fetchers
+        // I.e. this dictionary is quite small, does not grow up after service warm-up and does not require clean up
+        private readonly ConcurrentDictionary<Property, PropertyFetcher> propertyFetchers = new ConcurrentDictionary<Property, PropertyFetcher>();
+
+        internal ServiceBusDiagnosticsEventHandler(TelemetryConfiguration configuration)
+        {
+            this.telemetryClient = new TelemetryClient(configuration);
+        }
+
+        /// <summary>
+        /// Gets custom handler <see cref="HandleDiagnosticsEvent{TContext}"/>delegate.
+        /// </summary>
+        internal HandleDiagnosticsEvent<HashSet<string>> EventHandler => this.HandleEvent;
+
+        private void HandleEvent(KeyValuePair<string, object> evnt, DiagnosticListener diagnosticListener, HashSet<string> context)
+        {
+            // there is no guarantee IsEnabled was called in case there are multiple listeners in the system. i.e. we need to check it again
+            if (evnt.Key.EndsWith("Start", StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            Activity currentActivity = Activity.Current;
+
+            switch (evnt.Key)
+            {
+                case "Microsoft.Azure.ServiceBus.ProcessSession.Stop":
+                case "Microsoft.Azure.ServiceBus.Process.Stop":
+                    this.OnRequest(evnt.Key, evnt.Value, currentActivity);
+                    break;
+                case "Microsoft.Azure.ServiceBus.Exception":
+                    this.OnException(evnt.Key, evnt.Value);
+                    break;
+                default:
+                    this.OnDependency(evnt.Key, evnt.Value, currentActivity);
+                    break;
+            }
+        }
+
+        private void OnDependency(string name, object payload, Activity activity)
+        {
+            DependencyTelemetry telemetry = new DependencyTelemetry { Type = DependencyType };
+
+            this.SetCommonProperties(name, payload, activity, telemetry);
+
+            // Endpoint is URL of particular ServiceBus, e.g. sb://myservicebus.servicebus.windows.net/
+            telemetry.Data = this.FetchPayloadProperty<Uri>(name, EndpointPropertyName, payload)?.ToString();
+
+            // Queue/Topic name, e.g. myqueue/mytopic
+            telemetry.Target = this.FetchPayloadProperty<string>(name, EntityPropertyName, payload);
```

Comment:
According to [Application Insights data model](https://docs.microsoft.com/en-us/azure/application-insights/application-insights-data-model-dependency-telemetry#target) should identify the side of dependency call. It is also used to identify nodes in AppMap. Entity path can be the same for different SB accounts so it does not serve that purpose. Endpoint hostname works better here.

Reply:
alternatively we could combine endpoint and entity but that would result in multiple nodes for the same SB account on AppMap - it might not scale properly in some cases (i.e. entity can have high cardinality for some customers)

Reply:
Right, and Target being queueName fits perfectly fine into the side of dependency call. Combining it with endpoint makes node name not readable (too long to show). So we need more rules here: e.g. showing queue name only, but distinguishing nodes based on Target+Endpoint. That's where having Endpoint in Data field is more efficient than in custom dimensions

Reply:
I'm not sure if that's the right interpretation of Target and the documentation leaves some wiggle room. @SergeyKanzhelev thoughts?

Reply:
We have many constraints here: 
1. we need to collect useful data that and be able to efficiently query it and build UI based on it
2. we have rules around the data model

Data model for dependencies so far has been used for http and sql only. Some protocols may not fit into it as we see it here.
So if we can't satisfy both requirements, I prefer we update the rules, or change schema rather than sacrifice p1 (data and efficiency)